### PR TITLE
Backend: Added External Police SOS apis

### PR DESCRIPTION
**Pull request by shivendrashah**
Backend: Added External Police SOS apis

## Summary by CodeRabbit

* **New Features**
  * External SOS support (ERSS, GJ112) with frontend and dashboard triggers; SOS can be sent to external providers, store external reference IDs, and upload media when required.
  * New SOS APIs: fetch SOS details, trigger external SOS call, and ERSS status update; UI/ride responses include detailed SOS info and external status/history.
  * Emergency settings expose external SOS configuration to clients.

* **Chores**
  * DB schema extended with external SOS config, polling interval, and external reference/status columns.

### CODE REVIEW

**Comment by Magizhan** [comment]
🤖 **AI Code Review**

**Summary:** Large feature addition for external police SOS integration (ERSS/GJ112). Adds comprehensive API integration, config management, authentication, and domain logic for emergency SOS dispatch.

**Findings:**

**[Bug]** Missing error handling for SOS config validation:
- In `Dashboard.Sos.callExternalSOS`, when `sosConfig.triggerSource /= DRC.DASHBOARD`, an error is thrown
- However, in `UI.Sos.postSosCreate`, the check is `sosConfig.triggerSource == DRC.FRONTEND` but no error is thrown otherwise
- This could allow SOS creation when triggerSource is neither FRONTEND nor DASHBOARD

**[Security]** Credential handling in `buildExternalSOSDetails`:
- Function decrypts `person.mobileNumber` and defaults to "0000000000" if missing
- Sending a default/fake phone number to emergency services could be problematic
- Consider throwing an error or requiring valid mobile number for SOS

**[Performance]** Token caching strategy:
- ERSS and GJ112 auth modules cache tokens in Redis with key prefix from config
- Good approach, but consider adding metrics/logging for token refresh failures
- Token refresh logic has 60s buffer before expiry—ensure this doesn't cause race conditions under high load

**[Suggestion]** Base64 encoding in GJ112 auth:
- GJ112 requires Base64-encoded credentials (username/password)
- The encoding happens in `getNewToken` but the config stores encrypted values
- Consider documenting why Base64 is needed (API requirement) to avoid future confusion

**[Bug]** ERSS response validation:
- `validateERSSResponse` checks `isERSSSuccess` but HTTP 200 responses with `OPERATION_FAILURE` might not be caught by Servant error handling
- Ensure all ERSS API calls go through `validateERSSResponse` to prevent silent failures

**[Suggestion]** Error messages in interface conversions:
- `fromGJ112SOSRes` checks `responseCode == 200` to determine success
- If responseCode is missing, it defaults to `False` success—consider explicit error handling

**[Style]** Large shared-kernel diff:
- Adds 1000+ lines of new SOS integration code across multiple providers
- Well-structured with clear separation (ERSS vs GJ112), following existing patterns (SMS/Maps)
- Consider splitting into smaller PRs in future for easier review

**[Suggestion]** Testing:
- No test files included for the new SOS integration
- Consider adding unit tests for auth token caching, response validation, and error handling

Overall: This is a substantial feature with good code structure. Main concerns are error handling edge cases and missing validation for external API responses. Recommend testing with actual ERSS/GJ112 staging endpoints before production deployment.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Dashboard/Sos.hs [hunk 1]
```diff
@@ -49,3 +65,45 @@ convertSosStatus Domain.Types.Sos.NotResolved = API.Types.RiderPlatform.Manageme
...
+    mkSosReq :: DSos.Sos -> UISos.SosReq
+    mkSosReq sos =
+      UISos.SosReq
+        { flow = sos.flow,
+          rideId = Just sos.rideId,
+          isRideEnded = Nothing,
+          sendPNOnPostRideSOS = Nothing,
+          notifyAllContacts = Nothing,
+          customerLocation = Nothing
+        }
```
[the remark this answers is not in the document]
Comment (bytedex) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Dashboard/Sos.hs]: @ayush-297 We will need to populate the customer's location (using driver/customer location or from ride location)

File: Backend/app/rider-platform/rider-app/Main/spec/Storage/sos.yaml [hunk 2]
```diff
@@ -61,6 +64,10 @@ Sos:
       kvFunction: updateOneWithKV
       params: [entityType]
       where: id
+    updateExternalReferenceId:
```
Comment (bytedex) [inline Backend/app/rider-platform/rider-app/Main/spec/Storage/sos.yaml]: @TejaswiniT167 we can remove this extra updateExternalReferenceId already defined in this yaml later on

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 3]
```diff
@@ -392,6 +422,23 @@ uploadMedia sosId personId SOSVideoUploadReq {..} = do
...
+            whenJust phoneNumber $ \phoneNo -> do
+              let sosServiceType = flowToSOSService sosConfig.flow
+              merchantSvcCfg <-
+                QMSC.findByMerchantOpCityIdAndService person.merchantId person.merchantOperatingCityId (DMSC.SOSService sosServiceType)
+                  >>= fromMaybeM (MerchantServiceConfigNotFound person.merchantOperatingCityId.getId "SOS" (show sosServiceType))
+              case merchantSvcCfg.serviceConfig of
+                DMSC.SOSServiceConfig specificConfig -> do
+                  let fileName = "sos-" <> getId sosId <> "." <> fileExtension
+                  void $
+                    withTryCatch "PoliceSOS:uploadMedia" $
```
Comment (bytedex) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs]: @TejaswiniT167 can we have some logError or some logInfo for debugging in kibana 😅

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 4]
```diff
@@ -766,3 +904,144 @@ postSosUpdateState (mbPersonId, _) sosId UpdateStateReq {..} = do
...
+  mobileNo <- fmap (fromMaybe "0000000000") $ traverse decrypt person.mobileNumber
+  emailText <- traverse decrypt person.email
+  imeiText <- traverse decrypt person.imeiNumber
+  customerDisability <- runInReplica $ PDisability.findByPersonId person.id
+  let (lat, lon) =
+        if sosConfig.latLonRequired
+          then case req.customerLocation of
+            Just loc -> (loc.lat, loc.lon)
+            Nothing -> (fromMaybe 0.0 person.latestLat, fromMaybe 0.0 person.latestLon)
+          else (0.0, 0.0)
```
Comment (bytedex) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs]: @shivendrashah  here also we can add a proper check that if somehow from customer request body customerLocation is not present, we should fetch the driverLatLon from rideId (sharedLogic module func)

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 5]
```diff
@@ -766,3 +904,144 @@ postSosUpdateState (mbPersonId, _) sosId UpdateStateReq {..} = do
...
+  let dateTimeStr = T.pack $ formatTime defaultTimeLocale "%Y-%m-%d %H:%M:%S" now
```
Comment (bytedex) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs]: @shivendrashah should we convert from UTC to IST

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 6]
```diff
@@ -766,3 +904,144 @@ postSosUpdateState (mbPersonId, _) sosId UpdateStateReq {..} = do
...
+        latitude = lat,
+        longitude = lon,
+        speed = Nothing,
+        mobileNo = mobileNo,
+        imeiNo = imeiText,
+        gpsProvider = Nothing,
+        senderName = person.firstName,
+        address = Nothing,
+        gpsAccuracy = Nothing,
+        stateCode = Just "07",
```
Comment (bytedex) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs]: we can use stateCode from riderConfig instead of hardcoding it to "07"

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs (hunk 4 above)
Comment (ayush-297) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs]: Already fixed 👍

File: Backend/app/rider-platform/rider-app/Main/spec/Storage/RiderConfig.yaml [hunk 7]
```diff
@@ -214,6 +225,8 @@ RiderConfig:
     kafkaTopicName: Text
     staticCustomerIdThresholdDay: Maybe Day
     invoiceLogoUrl: Maybe Text
+    externalSOSConfig: Maybe ExternalSOSConfig
+    externalSOSTracePollingIntervalSeconds: Maybe Seconds
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/spec/Storage/RiderConfig.yaml]: Why is this separate from config?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 8]
```diff
@@ -392,6 +423,24 @@ uploadMedia sosId personId SOSVideoUploadReq {..} = do
...
+            -- Skip if no phone number available
+            whenJust phoneNumber $ \phoneNo -> do
+              let sosServiceType = flowToSOSService sosConfig.flow
+              merchantSvcCfg <-
+                QMSC.findByMerchantOpCityIdAndService person.merchantId person.merchantOperatingCityId (DMSC.SOSService sosServiceType)
+                  >>= fromMaybeM (MerchantServiceConfigNotFound person.merchantOperatingCityId.getId "SOS" (show sosServiceType))
+              case merchantSvcCfg.serviceConfig of
+                DMSC.SOSServiceConfig specificConfig -> do
+                  let fileName = "sos-" <> getId sosId <> "." <> fileExtension
+                  uploadMediaRes <- PoliceSOS.uploadMedia specificConfig phoneNo fileName payload
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs:439]: Better to use pre-signed url if media is large

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 9]
```diff
@@ -557,8 +606,102 @@ postSosUpdateLocation (mbPersonId, _) sosId req = do
...
+      --   Nothing -> 3600 :: Int
+  case cached of
+    Just (_, False, _) -> pure ()
+    Just (_, True, Nothing) -> pure ()
+    Just (lastTraceSec, True, Just specificConfig) ->
+      when (nowSec - lastTraceSec >= fromIntegral pollingIntervalSec) $
+        whenJust sosDetails.externalReferenceId $ \trackingId -> do
+          mbMobile <- getPersonMobileNo sosDetails.personId
+          whenJust mbMobile $ \mobileNo -> do
+            callSOSTraceAPI trackingId specificConfig mobileNo req timeDiff
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs:634]: To be discussed later: Frontend to have all these logics

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs [hunk 10]
```diff
@@ -766,3 +912,160 @@ postSosUpdateState (mbPersonId, _) sosId UpdateStateReq {..} = do
...
+        email = emailText,
+        vendorName = Just merchantOpCity.merchantShortId.getShortId,
+        deviceType = Just 2
+      }
+
+resolveLocation :: Maybe Maps.LatLong -> Person.Person -> Maybe DRide.Ride -> Flow (Double, Double)
+resolveLocation mbCustomerLoc person mbRide =
+  case mbCustomerLoc of
+    Just loc -> pure (loc.lat, loc.lon)
+    Nothing -> case (person.latestLat, person.latestLon) of
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs:1019]: To send location type also

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs (hunk 9 above)
Comment (bytedex) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs:634]: Will do this in a separate enh pr @shivendrashah, when moving this logic to LTS.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs (hunk 10 above)
Comment (bytedex) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Sos.hs:1019]: Will do this in a separate enh pr @shivendrashah, when moving this logic to LTS.
